core/src/test/java/org/apache/iceberg/TestWapWorkflow.java

```diff
@@ -684,4 +685,87 @@ public void testNonWapCherrypick() {
           table.manageSnapshots().cherrypick(firstSnapshotId).commit();
         });
   }
+
+  @Test
+  public void testWithExpiringWapThenCherrypick() {
+    table.newAppend()
+            .appendFile(FILE_A)
+            .commit();
+    // first WAP commit
+    table.newAppend()
+            .appendFile(FILE_B)
+            .set(SnapshotSummary.STAGED_WAP_ID_PROP, "123456789")
+            .stageOnly()
+            .commit();
+    table.refresh();
+
+    // pick the snapshot that's staged but not committed
+    Snapshot wap1Snapshot = null;
+    for (Snapshot s : table.snapshots()) {
+      if ("123456789".equalsIgnoreCase(s.summary()
+              .getOrDefault(SnapshotSummary.STAGED_WAP_ID_PROP, null))) {
+        wap1Snapshot = s;
+        break;
+      }
+    }
+    Assert.assertNotNull(wap1Snapshot);
+
+    // table has new commit so that cherrypick can come into affect.
+    table.newAppend()
+            .appendFile(FILE_C)
+            .commit();
+    table.refresh();
+
+    // cherry-pick first snapshot
+    table.manageSnapshots().cherrypick(wap1Snapshot.snapshotId()).commit();
+    table.refresh();
+    Snapshot wap1CherryPickSnapshot = table.currentSnapshot();
+    Assert.assertEquals("123456789", wap1CherryPickSnapshot.summary().getOrDefault(SnapshotSummary.PUBLISHED_WAP_ID_PROP,
+            null));
+    Assert.assertEquals(wap1CherryPickSnapshot.summary().getOrDefault(SnapshotSummary.SOURCE_SNAPSHOT_ID_PROP,
+            null), String.valueOf(wap1Snapshot.snapshotId()));
+
+
+    table.expireSnapshots().expireOlderThan(wap1Snapshot.timestampMillis() + 1).commit();
+    table.expireSnapshots().expireOlderThan(wap1CherryPickSnapshot.timestampMillis() + 1).commit();
+
+    //Make sure no files are deleted for the wap and cherrypick
+    Lists.newArrayList(wap1Snapshot, wap1CherryPickSnapshot).forEach(i -> {
+      i.addedFiles().forEach(item -> {
+        Assert.assertFalse(TestTables.DELETED_FILE_PATHS.contains(item.path().toString()));
+      });
+    });
+  }
+
+  @Test
+  public void testWithExpiringDanglingWap() {
```

Comment:
This UT fails currently, but will start to work fine, once we rebase with latest master for fixes merged via PR https://github.com/apache/incubator-iceberg/pull/812